CONTRIBUTING.md

```diff
@@ -0,0 +1,338 @@
+# Development Conventions
+
+Welcome to the Executable Book Project (EBP)!
+We're excited you're here and want to contribute.
+
+This page outlines conventions and best practices for development and maintenance across all repositories in the EBP organisation, to help the community make the best tools possible.
+
+Thank you for you interest in contributing ✨
+
+(dev/code_style)=
+
+## Coding Style
+
+Coding style is largely enforced automatically, using [pre-commit hooks](https://pre-commit.com/).
+For Python packages, the pre-commit should include automated code formatting *via* [Black](https://black.readthedocs.io/) and code linting *via* [flake8](https://flake8.pycqa.org).
+
+(dev/testing)=
+
+## Testing
+
+All packages should contain a test infrastructure, usually automated for PRs and commits *via* [GitHub Actions workflows](https://docs.github.com/en/actions/configuring-and-managing-workflows/configuring-a-workflow).
+
+Testing philosophy:
+
+- Whenever you encounter a bug, add a (failing) test for it. Then fix the bug.
+- Whenever you modify or add a feature, write a test for it!
+  Writing tests [before writing the actual implementation](http://en.wikipedia.org/wiki/Test_Driven_Development) helps keeping your API clean.
+- Make [unit tests](https://en.wikipedia.org/wiki/Software_testing#Testing_levels) as atomic as possible. A unit test should run in the blink of an eye.
+- Document why you wrote your test - developers will thank you for it once it fails.
+
+For Python packages, the test infrastructure should be implemented *via* [pytest](https://docs.pytest.org).
+
+(dev/docs)=
+
+## Documentation
+
+A minimal description of the project should be contained in the README.md, then most documentation should generally be contained in a `docs` folder, using [Sphinx](http://www.sphinx-doc.org) (directly or *via* jupyter book) as the documentation generator.
+
+Markdown style should generally follow that rules outlined [markdownlint](https://github.com/markdownlint/markdownlint) and rST should similarly follow [rstcheck](https://github.com/myint/rstcheck), either of which may be added to the pre-commit configuration.
+
+In addition, when writing documentation authors should adhere to the following guidelines:
+
+1. Write **one sentence per line** and otherwise **no manual line wrapping** to make easy to create and review  diffs. All standard editors allow for dynamic line wrapping, and the line length is irrelevant for the rendered documentation in, e.g., HTML or PDF format.
```

Comment:
hmmm, this one would mean that almost all of our docs are in violation of this 😬 - I can see an argument both ways. I'm assuming this is copypasta but what do you think?

Reply:
Well its copied from https://github.com/aiidateam/aiida-core/wiki/Writing-documentation#style-guide, but it is something I agree with and try to adhere to and enforce.
Its not a hard and fast rule (e.g. we aren't going to reject a PR over it) but it is something I want to encourage.

Reply:
that makes sense - to be honest I'd never thought about it before, I wonder if I can make my editors soft-wrap at a particular line length even if the window is wider than needed?

Reply:
🤷 in vscode I use in settings.json:

```json
 "[markdown]": {
        "editor.wordWrap": "on",
    }
```